crates/polars-core/src/datatypes/any_value.rs

```diff
@@ -447,38 +455,40 @@ impl<'a> AnyValue<'a> {
         macro_rules! cast_to (
             ($av:expr) => {
                 match dtype {
-                    DataType::UInt8 => AnyValue::UInt8($av as u8),
-                    DataType::UInt16 => AnyValue::UInt16($av as u16),
-                    DataType::UInt32 => AnyValue::UInt32($av as u32),
-                    DataType::UInt64 => AnyValue::UInt64($av as u64),
-                    DataType::Int8 => AnyValue::Int8($av as i8),
-                    DataType::Int16 => AnyValue::Int16($av as i16),
-                    DataType::Int32 => AnyValue::Int32($av as i32),
-                    DataType::Int64 => AnyValue::Int64($av as i64),
-                    DataType::Float32 => AnyValue::Float32($av as f32),
-                    DataType::Float64 => AnyValue::Float64($av as f64),
+                    DataType::UInt8 => AnyValue::UInt8($av.try_extract::<u8>()?),
```

Comment:
Hang on, the function returns a `None` on a failed cast regardless, the strictness of the cast determines whether or not this propagates into a `null` value or whether an error is thrown, and that is dealt elsewhere. The new implementation respects the `strict` parameter in the same way that it does for `Series` and `Expr`. Prior to this, `pl.lit.cast(strict=False)` acted differently than `Series.cast` and `Expr.cast`, both of which would return null instead of 255.

Reply:
Do we mean propagating the error instead of None when `strict=True`?

Reply:
Also @nameexhaustion the `DataType::Int64 => AnyValue::Int64(*v as i64)` is guaranteed not to error, that's just extracting the physical value from the `Datetime`, ditto for the `Time` and `Date` below.

Reply:
Sorry, to clarify. In the old implementation, this function returned an *actual value* regardless of the strictness parameter. This behavior differs from how series and expressions treated the `strict=False` paramter, which would return `None`.

In the new implementation, this function acts as if always strict and errors when it's an invalid cast. That error is picked up in the `simplify_exprs` clause and converted to a None. Are you saying that logic (to return None) should be pushed into the `Anyvalue` cast with a strict/non-strict? That's fine with me.

Reply:
I can still clean this up a bit. I'm trying to figure out how to trigger some of these codepaths (the int/floats seem to always get hit, even with logical dtypes), and I'll work on pushing strict into `any_value.rs`.

Reply:
Ok @ritchie46 and @nameexhaustion, I've pushed a new commit that I think is much cleaner and resolves most of our qualms:

* `any_value.cast` now takes the `strict` parameter. When the cast fails and `strict=False`, the `None` value is returned. This is a change in behavior from before where it would return a UB value, but that old behavior did not align with how `strict=False` worked for series and expressions. The new behavior does align and returns null. When the cast fails and `strict=True`, a `ComputeError` is raised.
* The casting is more robust, i.e. supports more casts than before, in an attempt to align with `Series.cast` and `Expr.cast`. For example, you can now perform `pl.lit(date(1970, 1, 1)).cast(pl.Int64)`.
* Added path for `pl.lit(x, p l.Duration).cast` which was previously absent.